blocks/api/parser.js

```diff
@@ -88,6 +53,69 @@ export function asType( value, type ) {
 }
 
 /**
+ * Returns an hpq matcher given a source object
+ *
+ * @param  {Object}   sourceConfig Attribute Source object
+ * @return {Function}              hpq Matcher
+ */
+export function matcherFromSource( sourceConfig ) {
+	switch ( sourceConfig.source ) {
+		case 'attribute':
+			return attr( sourceConfig.selector, sourceConfig.attribute );
+		case 'property':
+			return prop( sourceConfig.selector, sourceConfig.property );
+		case 'html':
+			return html( sourceConfig.selector );
+		case 'text':
+			return text( sourceConfig.selector );
+		case 'children':
+			return children( sourceConfig.selector );
+		case 'node':
+			return node( sourceConfig.selector );
+		case 'query':
+			return query( sourceConfig.selector, matcherFromSource( sourceConfig.query ) );
+		case 'object':
+			return keys( sourceConfig.object ).reduce( ( memo, key ) => {
+				memo[ key ] = matcherFromSource( sourceConfig.object[ key ] );
+				return memo;
+			}, {} );
+		default:
+			// eslint-disable-next-line no-console
+			console.error( `Unkown source type "${ sourceConfig.source }"` );
+	}
+}
+
+/**
+ * Given a blocktype, a block's raw content and the commentAttributes returns the attribute value depending on its source definition
+ *
+ * @param  {string} attributeKey        Attribute key
+ * @param  {Object} attributeSchema     Attribute's schema
+ * @param  {string} rawContent          Block's raw content
+ * @param  {Object} commentAttributes   Block's comment attributes
+ *
+ * @return {mixed}                      Attribute value
+ */
+export function getBlockAttribute( attributeKey, attributeSchema, rawContent, commentAttributes ) {
+	let value;
+	switch ( attributeSchema.source ) {
+		case 'meta':
+			break;
+		case 'comment':
```

Comment:
Hmm, we've tried to avoid being explicit about comment as a source, treating it as a default. Here instead we treat default as though source is one of the hpq matchers. Maybe instead we should include the cases for those sources in `matcherFromSource` and let comment be the default case?

Reply:
> Hmm, we've tried to avoid being explicit about comment as a source

Why? I see value in leaving the matchers as the non-explicit ones to avoid listing all the matchers.

It makes sense to me to have a clear `matcherFromSource` function responsible only for sourcing attributes from the content.

Reply:
>Why? I see value in leaving the matchers as the non-explicit ones to avoid listing all the matchers.

Implementation aside, it's if we conceptually consider serialized comments to be the default storage for attributes, but support other strategies for sourcing these values.

Reply:
To this point, we ought to consider consistency with how we refer to the `commentAttributes` argument here. In `createBlockWithFallback`, this is named `attributes`. In `parseWithGrammar`, the block node property is named `attrs`.

Reply:
Do you think I should use `attrs` in all these places?

Reply:
I'm not a huge fan of abbreviations. If we were to think of comment attributes as the "base" attribute set, then `attributes` would be appropriate.

Reply:
it's confusing to me. The function is called `getBlockAttributes`, it shouldn't have `attributes` as an argument don't you think?

Reply:
Yes, this is confusing, I agree. This ties into disjointedness of our attribute sourcing generally (https://github.com/WordPress/gutenberg/pull/2854/files#r144836758). What are these attributes which are magically prepopulated? Are the comments a core part of what a block is, or an incidental implementation detail? In the current context, I could see `parsedAttributes` making sense, but our approach of passing content and comment attributes as arguments doesn't really scale well to other source types, and as implemented we are aware of all but implementing behaviors of only some types. 

How capable do we want the parsing to be? I don't necessarily think we need to build in concepts of post meta and site options to the parser itself, but maybe as some extensibility pattern: either allowing additional sources to inject their values to be referenced by the parser when encountering that source type, or hooks to allow the post editor a turn when encountering the parsing or serialization behaviors of a specific source type.

Alternatively, if it's the case that our editor selector is responsible for managing the implementation of specific types ([post meta](https://github.com/WordPress/gutenberg/blob/1b68a192410eaeba5e494f6f4334045b0bc65a5a/editor/selectors.js#L435-L482)), why should this not also extend to sourcing from content as well? Ultimately maybe there is a line where a blocks core knows to handle serialized and markup-sourced attributes, then allow additional handlers to be reflected later. But ideally not with built-in awareness of what those other handlers are; if we dropped `case 'meta'` and treated comments as the default, would it work just as well?

Reply:
>  if we dropped case 'meta' and treated comments as the default, would it work just as well?

Can you provide a code example, I don't understand what you're proposing here.

> : either allowing additional sources to inject their values to be referenced by the parser when encountering that source type,

This won't work for site options and post meta because their value change over time, and using the parser for this is not a good option. The parser is meant for parsing post_content (HTML) and the content contains only two types of attributes: comment attributes and attributes parsed from HTML.

>  hooks to allow the post editor a turn when encountering the parsing or serialization behaviors of a specific source type.

Maybe, this means hooks to the `getBlock` selcetor and `updateAttributes` callback. 

---

I think we're struggling here because we're trying to fix the symptoms instead of trying to fix the root issue. And the root issue is the Editor not being generic enough. Things would much clearer if the Editor was a generic content editor (not tied to post or site) with extensibility mechanisms. 

I do think this is out of scope of the current PR

Reply:
>I think we're struggling here because we're trying to fix the symptoms instead of trying to fix the root issue. And the root issue is the Editor not being generic enough. Things would much clearer if the Editor was a generic content editor (not tied to post or site) with extensibility mechanisms.

This is really what I'm trying to get at in my earlier comment, and arguably it is in scope because we're introducing knowledge of a meta type to the parser with these changes that had not existed previously.